bip-ccv.mediawiki

```diff
@@ -0,0 +1,416 @@
+
+-----
+
+::[[File:bip-ccv/amount_example_1.png|framed|center|alt=1-to-1 amount logic|600px]]
+::'''Figure 1:''' A UTXO <code>A</code> sends the entire amount to some output contract <code>B</code>, using <code>CCV</code> with the <i>default</i> semantic.
+
+-----
+
+::[[File:bip-ccv/amount_example_2.png|framed|center|alt=3-to-1 aggregate amount logic|600px]]
+::'''Figure 2:''' Three UTXOs aggregate their amounts towards the same output contract, using <code>CCV</code> with the <i>default</i> semantic.
+
+-----
+
+::[[File:bip-ccv/amount_example_3.png|framed|center|alt=split amount logic|600px]]
+::'''Figure 3:''' A UTXO <code>A</code> sends a portion of its amount to a contract <code>A'</code> identical to itself, and the rest to a different contract <code>B</code>. It would use <code>CCV</code> to introspect its own input's program, then to check the first output with the <i>deduct</i> semantic, then to check the second output with the <i>default</i> semantic to assign the residual amount.
+
+-----
+
+::[[File:bip-ccv/amount_example_4.png|framed|center|alt=split and aggregate amount logic|600px]]
+::'''Figure 4:''' Similar to the previous example, but a second input <code>B</code> also checks the same output <code>X</code> with the <i>default</i> semantic, aggregating its input with the residual amount of the first input.
+
+-----
+
+Note that the ''deduct'' semantic does not allow to check the exact amount of its output. Therefore, in contracts using a scheme similar to figure 3 or 4 above, amounts be constrained with a signature, or with future introspection opcodes that allow fixing the amount.
+
+== Specification ==
+
+The tapscript opcode <code>OP_SUCCESS187</code> (<code>0xbb</code>) is constrained with new rules to implement <code>OP_CHECKCONTRACTVERIFY</code>.
+
+When evaluating <code>OP_CHECKCONTRACTVERIFY</code> (<code>OP_SUCCESS187</code>,
+<code>0xbb</code>), the expected format of the stack, shown top to bottom, is:
+
+<source>
+<flags>
+<taptree>
+<pk>
+<index>
+<data>
+</source>
+
+where:
+
+* <code><flags></code> is a minimally encoded integer, according to one of the values defined below.
+* <code><taptree></code> is the Merkle root of the taproot tree, or a minimally encoded <code>-1</code>, or the empty buffer.
+* <code><pk></code> is called the ''naked key'', and it's a valid 32-byte x-only public key, or a minimally encoded <code>-1</code>, or the empty buffer.
+* <code><index></code> is a minimally encoded -1, or a minimally encoded non-negative integer.
+* <code><data></code> is a buffer of arbitrary length.
+
+In short, the semantics of the opcode with respect to the Script can be summarized as follows:
+
+  Verify that the input/output with the given <code>index</code> is a P2TR output where the public key is obtained from <code><pk></code>, tweaked with the hash of <data> (if non-empty), then taptweaked with <code><taptree</code> (if present).
```

Comment:
This [renders](https://github.com/Merkleize/bips/blob/ccv/bip-ccv.mediawiki#specification) as a code block without wrapping ⇒ have to scroll to the right to read it. Plain text (or blockquote, if you want to emphasize) might be better.

Reply:
b2199950
I added the blockquote, and rephrased a bit - saying P2TR UTXO instead of the ambiguous 'output'. Does this make it clearer?